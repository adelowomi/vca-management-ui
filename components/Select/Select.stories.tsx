import { action } from '@storybook/addon-actions';
import { boolean, text, withKnobs } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import React from 'react';

import Select from './Select';

const data = [
  { id: 1, value: 'Governmant', label: 'Governmant' },
  { id: 2, value: 'Tech', label: 'Tech' },
  { id: 3, value: 'Law', label: 'Law' },
];
storiesOf('Select', module)
  .addDecorator(withKnobs)
  .add('Select Demo', () => (
    <Select
      label={text('Label', 'Industry')}
      placeholder={text('Placeholder', 'Placeholder')}
      options={data}
      onSelect={action('onSelect')}
      disabled={boolean('Disabled', false)}
    />
  ))
  .add('Disabled Select', () => (
    <Select
      label={text('Label', 'Industry')}
      placeholder={text('Placeholder', 'Placeholder')}
      options={data}
      onSelect={action('onSelect')}
      disabled={true}
    />
  ));
