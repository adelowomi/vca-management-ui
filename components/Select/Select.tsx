import React, { useCallback } from 'react';

import mergeClassNames from '../../helpers/mergeClassNames';
import { ChevronDown, ChevronUp } from '../AssetsSVG';

interface Option {
  value: string;
  label: string;
}

export interface SelectProps {
  /**
   * Add some additional styles
   */
  className?: string;
  /**
   * Boolean of whether to make the input disabled
   */
  disabled?: boolean;
  /**
   * Determines label of select field
   */
  label: string;
  /**
   * Add some additional styles to the list
   */
  listClassName?: string;
  /**
   * specifies a short hint that describes the expected value
   * in the select field
   */
  placeholder: string;
  /**
   * List of items to display in the dropdown
   */
  options: Option[];
  /**
   * Callback for the consuming application
   */
  onSelect: Function;
}

const Select: React.FC<SelectProps> = ({
  className,
  disabled,
  label,
  placeholder,
  listClassName,
  options,
  onSelect,
}) => {
  const [show, setShow] = React.useState(false);
  const [selectedValue, setSelectedValue] = React.useState('');

  const toggleOpen = useCallback(() => {
    if (!disabled) setShow((val) => !val);
  }, [disabled]);

  const selectThis = useCallback(
    (option: Option) => (event: React.MouseEvent | React.KeyboardEvent) => {
      event.stopPropagation();
      onSelect(option.label);
      setSelectedValue(option.label);
      setShow(false);
    },
    [onSelect, setShow, setSelectedValue]
  );

  const handleClickOutside = () => {
    setShow(false);
  };

  const renderListItems = (items: Option[]) =>
    items.map((item, index) => {
      return (
        <li
          id={`${item.value}`}
          key={index}
          role="option"
          onKeyPress={selectThis(item)}
          onClick={selectThis(item)}
          data-testid={'selectList'}
        >
          {item.label}
        </li>
      );
    });
  const listStyle = show ? { display: 'block' } : { display: 'none' };
  const disabledClassName = disabled && 'app-select-disabled';

  return (
    <div className="flex flex-col w-full py-2 px-0">
      <label className="mb-2 font-medium app-greyish-brown">{label}</label>
      <div
        className={mergeClassNames(
          'relative inline-block',
          disabledClassName,
          className
        )}
        onClick={toggleOpen}
        onBlur={handleClickOutside}
        onKeyPress={toggleOpen}
        tabIndex={0}
        role="button"
        data-testid="select-container-test"
      >
        <div className={`app-form-select ${disabledClassName}`}>
          <div className="py-4">
            {selectedValue ? selectedValue : placeholder}
          </div>
        </div>
        <div className="absolute top-4 right-2 cursor-pointer">
          {show ? ChevronUp : ChevronDown}
        </div>
        <ul
          className={mergeClassNames('app-form-select-list', listClassName)}
          style={listStyle}
          role="listbox"
          data-testid="select-list-test"
        >
          {renderListItems(options)}
        </ul>
      </div>
    </div>
  );
};

const noOp = () => undefined;

Select.defaultProps = {
  className: '',
  disabled: false,
  label: '',
  placeholder: '',
  listClassName: '',
  options: [],
  onSelect: noOp,
};

export default Select;
