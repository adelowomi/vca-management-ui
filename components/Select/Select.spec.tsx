import '@testing-library/jest-dom';
import '@testing-library/jest-dom/extend-expect';

import { cleanup, fireEvent, render } from '@testing-library/react';
import React from 'react';

// Components
import Select from './Select';

afterEach(cleanup);

describe('Select', () => {
  const data = [
    { id: 1, value: 'Governmant', label: 'Governmant' },
    { id: 2, value: 'Tech', label: 'Tech' },
    { id: 3, value: 'Law', label: 'Law' },
  ];
  test('render properly', () => {
    const mockCallBack = jest.fn();
    const { container } = render(
      <Select
        label="Industry"
        placeholder={'Select'}
        options={data}
        onSelect={mockCallBack}
      />
    );
    expect(container).toMatchSnapshot();
  });
  describe('Select a Government from dropdown', () => {
    test('trigers a clcik event that selects Government', () => {
      const mockCallBack = jest.fn();
      const { getByText } = render(
        <Select
          onSelect={mockCallBack}
          label="Industry"
          placeholder={'Select'}
          options={data}
        />
      );
      fireEvent.click(getByText(/Governmant/i));
      expect(mockCallBack).toBeCalledWith('Governmant');
      expect(mockCallBack).toHaveBeenCalled();
    });
  });
  describe('Select is in disabled state', () => {
    test('render a disabled Select', () => {
      const mockCallBack = jest.fn();
      const { getByTestId, getByText } = render(
        <Select
          label="Industry"
          placeholder={'Select'}
          options={data}
          onSelect={mockCallBack}
          disabled={true}
        />
      );
      fireEvent.click(getByText(/Industry/i));
      expect(getByTestId('select-container-test')).toHaveClass(
        'app-select-disabled'
      );
      expect(getByTestId('select-list-test')).toHaveAttribute(
        'style',
        'display: none;'
      );
      expect(mockCallBack).not.toHaveBeenCalled();
    });
  });
});
