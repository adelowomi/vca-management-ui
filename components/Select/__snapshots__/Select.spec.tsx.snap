// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select render properly 1`] = `
<div>
  <div
    class="flex flex-col w-full py-2 px-0"
  >
    <label
      class="mb-2 font-medium app-greyish-brown"
    >
      Industry
    </label>
    <div
      class="relative inline-block"
      data-testid="select-container-test"
      role="button"
      tabindex="0"
    >
      <div
        class="app-form-select false"
      >
        <div
          class="py-4"
        >
          Select
        </div>
      </div>
      <div
        class="absolute top-4 right-2 cursor-pointer"
      >
        <svg
          height="30"
          version="1.1"
          viewBox="250 250 300 300"
          width="30"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title />
          <g
            id="icomoon-ignore"
          />
          <path
            d="M311.166 359.1l65.155 65.155c4.246 4.247 11.12 4.247 15.355 0l65.156-65.156c4.247-4.246 4.247-11.12 0-15.355s-11.12-4.246-15.355 0l-57.479 57.478-57.478-57.478c-4.246-4.247-11.12-4.247-15.355 0s-4.246 11.12 0 15.355z"
            fill="#0c1a38"
          />
        </svg>
      </div>
      <ul
        class="app-form-select-list"
        data-testid="select-list-test"
        role="listbox"
        style="display: none;"
      >
        <li
          data-testid="selectList"
          id="Governmant"
          role="option"
        >
          Governmant
        </li>
        <li
          data-testid="selectList"
          id="Tech"
          role="option"
        >
          Tech
        </li>
        <li
          data-testid="selectList"
          id="Law"
          role="option"
        >
          Law
        </li>
      </ul>
    </div>
  </div>
</div>
`;
